### ISSUE TITLE
Adding RegExp support on the color rules

### ISSUE THREAD

**Issue by burnersk** [burnersk/Dancer2-Logger-Console-Colored#1]
Feature request from @ambs

> Are you planning on adding regexp support on the color rules?
> 
> It would be nice to have 404 (routes not existing) appear in red or
> something like that...

**Comment by ambs** [comment]
weeeeee.

**Comment by burnersk** [comment]
@ambs it's a *little* delayed but here it is ;)

**Comment by simbabque** [comment]
@ambs I realize I'm late to the party, but I was just looking into the module to fix a bug and I think we can add this, too. Did you have anything in mind in terms of interface two years ago? I have a vague idea of how this could work, but don't see how exactly that would be useful.

**Comment by ambs** [comment]
Something like:

``````
    colored_regex:
      - re: /404/
        color: 'some_color_definition'
      - re: /missing/
        color: 'some_other_color'
``````````
Just an option. Given @burnersk is the original author, maybe he has other preferences.

**Comment by simbabque** [comment]
@ambs I'm the current maintainer. Unfortunately @burnersk is not doing a lot of Perl any more. So you're stuck with me. ;)

I do like this interface. Not sure about the `/` though. I don't think YAML or JSON are smart enough to convert those to `qr//`. A regular string is probably easier to parse, but it could still be treated as a pattern. It would just have to be documented.

> colored_regex:
>   - re: 404
>     color: 'some_color_definition'
>   - re: (?:missing|not there)
>     color: 'some_other_color'

What do you think about that?

Here are some more questions for the behavior.

- Would we match the regex only on the log message? 
- What would happen when this regex matches? 
- Would just the match be colored, or the whole line?

**Comment by ambs** [comment]
Heh, sorry. Good someone is maintaining it :-)

Yes, the bars are not relevant. My bad.

Answering your questions:
 - I would match the whole printed line 
 - regex would have precedence over type-colors
 - color the match (if you want the whole like, just make the re a while line)

Those are just my "quick" ideas. Be free to stick with your preferences.

**Comment by simbabque** [comment]
How do you feel about this? It uses the interface you suggested but for the last point.

![image](https://user-images.githubusercontent.com/1125067/28181471-0b8e415c-6809-11e7-9dab-44f28dbd1a04.png)

> color the match (if you want the whole like, just make the re a while line)

I think you meant _if you want the whole line, make the re a white line_. I'm not sure I understand, though. I'd just use `.+` to color the whole line. That would look like this:

![image](https://user-images.githubusercontent.com/1125067/28181575-69ec1a94-6809-11e7-9e08-abc197229005.png)

It allows to match more than one pattern as you can see from the first screenshot. After it matched, it will not color the message with the regular (default or custom) colors. The level will still be colored.

I've pushed [a branch](https://github.com/burnersk/Dancer2-Logger-Console-Colored/tree/gh-1-colored-regex) that has this change, but no tests or documentation yet. Feel free to play around with it and give feedback. :)

**Comment by ambs** [comment]
In fact, I wouldn't mind about what users do with it. If they want a rainbow, or if they want all digits in other color, it is up to them.

What I mean about matching the whole line was to use `re = .+foo.+` if you want all lines with foo colored somehow.

Also, I would first apply the default rules and then rewrite it using the RE rules.

I won't be able to test atm, sorry for that, but at the moment I am struggling with other things apart from Perl.

**Comment by simbabque** [comment]
Thanks for taking the time to get back to me so quickly. I'll see what I can do with the recursive coloring. Matching that properly without breaking previously applied coloring will be hard I guess.

**Comment by ambs** [comment]
Yeah, probably. If it seems over complicated, just ignore my ideas :)

**Comment by simbabque** [comment]
I've just merged my implementation and released version 0.004 on CPAN. It does not do recursive, but I think it's pretty useful nonetheless. Let me know if there is anything else you want this to do.
